darts/utils/optimised_historical_forecasts.py

```diff
@@ -0,0 +1,427 @@
+
+    For TimeSeries with a RangeIndex, the boundaries are converted to absolute indexes to slice the array appropriately
+    when start > 0.
+
+    When applicable, move the start boundaries to the value provided by the user.
+    """
+    # obtain forecastable indexes boundaries, as values from the time index
+    historical_forecasts_time_index = model._get_historical_forecastable_time_index(
+        series,
+        past_covariates,
+        future_covariates,
+        is_training=False,
+        reduce_to_bounds=True,
+    )
+
+    if historical_forecasts_time_index is None:
+        raise_log(
+            ValueError(
+                "Cannot build a single input for prediction with the provided model, "
+                f"`series` and `*_covariates` at series index: {series_idx}. The minimum "
+                "prediction input time index requirements were not met. "
+                "Please check the time index of `series` and `*_covariates`."
+            )
+        )
+        return ()
+
+    # shift the end of the forecastable index based on `overlap_end`` and `forecast_horizon``
+    last_valid_pred_time = model._get_last_prediction_time(
+        series,
+        forecast_horizon,
+        overlap_end,
+    )
+
+    historical_forecasts_time_index = (
+        historical_forecasts_time_index[0],
+        min(historical_forecasts_time_index[1], last_valid_pred_time),
+    )
+
+    # when applicable, shift the start of the forecastable index based on `start`
+    if start is not None:
+        start_time_ = series.get_timestamp_at_point(start)
+        # ignore user-defined `start`
+        if (
+            not historical_forecasts_time_index[0]
+            <= start_time_
+            <= historical_forecasts_time_index[-1]
+        ):
+            if show_warnings:
+                model._historical_forecasts_start_warnings(
+                    idx=series_idx,
+                    start=start,
+                    start_time_=start_time_,
+                    historical_forecasts_time_index=historical_forecasts_time_index,
+                )
+        else:
+            historical_forecasts_time_index = (
+                max(historical_forecasts_time_index[0], start_time_),
+                historical_forecasts_time_index[1],
+            )
+
+    # re-adjust the slicing indexes to account for the lags
+    (
+        min_target_lag,
+        _,
+        min_past_cov_lag,
+        _,
+        min_future_cov_lag,
+        max_future_cov_lag,
+    ) = model.extreme_lags
+
+    # target lags are <= 0
+    hist_fct_tgt_start, hist_fct_tgt_end = historical_forecasts_time_index
+    if min_target_lag is not None:
+        hist_fct_tgt_start += min_target_lag * freq
+    hist_fct_tgt_end -= 1 * freq
+    # past lags are <= 0
+    hist_fct_pc_start, hist_fct_pc_end = historical_forecasts_time_index
+    if min_past_cov_lag is not None:
+        hist_fct_pc_start += min_past_cov_lag * freq
```

Comment:
Since `historical_forecasts_time_index` already corresponds to the "forecastable target" indexes, we could leave all the work to the tabularization module and pass the entire covariates series. I wanted to minimize the length of these series while making sure that tabularization had enough points to work with:
- push back the start of past covariates if some lags are defined
- push back the start of future covariates if some lags are negative
- push forward the end of future covariates if some lags are positive

I am going to push around the boundaries in all scenario, for extensiveness but I don't think that it'll make a big difference